lib/rspec/core/bisect/example_minimizer.rb

```diff
@@ -0,0 +1,125 @@
+RSpec::Support.require_rspec_core "bisect/subset_enumerator"
+
+module RSpec
+  module Core
+    module Bisect
+      # @private
+      # Contains the core bisect logic. Searches for examples we can ignore by
+      # repeatedly running different subsets of the suite.
+      class ExampleMinimizer
+        attr_reader :runner, :reporter, :all_example_ids, :failed_example_ids
+
+        def initialize(runner, reporter)
+          @runner   = runner
+          @reporter = reporter
+        end
+
+        def find_minimal_repro
+          prep
+
+          remaining_ids = all_example_ids - failed_example_ids
+          debug 0, "Initial failed_example_ids: #{failed_example_ids}"
+          debug 0, "Initial remaining_ids: #{remaining_ids}"
+
+          each_bisect_round(lambda { remaining_ids }) do |subsets|
+            ids_to_ignore = subsets.find do |ids|
+              get_same_failures?(remaining_ids - ids)
+            end
+
+            next :done unless ids_to_ignore
+
+            remaining_ids -= ids_to_ignore
+            debug 1, "Removed #{ids_to_ignore}; remaining_ids: #{remaining_ids}"
+          end
+
+          remaining_ids + failed_example_ids
+        end
+
+      private
+
+        def prep
+          notify(:bisect_starting, :original_cli_args => runner.original_cli_args)
+
+          _, duration = track_duration do
+            original_results    = runner.original_results
+            @all_example_ids    = original_results.all_example_ids
+            @failed_example_ids = original_results.failed_example_ids
+          end
+
+          notify(:original_bisect_run_complete, :failures => failed_example_ids.size,
+                                                :non_failures => non_failing_example_ids.size,
+                                                :duration => duration)
+        end
+
+        def non_failing_example_ids
+          @non_failing_example_ids ||= all_example_ids - failed_example_ids
+        end
+
+        def get_same_failures?(ids)
+          results = runner.run(ids + failed_example_ids)
+          notify(:individual_run_complete)
+
+          abort_if_ordering_inconsistent(results)
+
+          (results.failed_example_ids == failed_example_ids).tap do |same|
+            if same
+              debug 2, "Running with #{ids}, got same failures"
+            else
+              debug 2, "Running with #{ids}, got different failures: #{results.failed_example_ids}"
+            end
+          end
+        end
+
+        def debug(level, msg)
+          puts "#{'  ' * level}Minimizer: #{msg}" if ENV['DEBUG_RSPEC_BISECT']
+        end
```

Comment:
> Is this intended for "release" usage? or just whilst this is WIP?

Kinda both, actually.  It was originally inspired by a similar feature in minitest-bisect:

https://github.com/seattlerb/minitest-bisect/blob/233b4fe69131358efa03da79270f91b2e7da9ea5/lib/minitest/find_minimal_combination.rb#L82-L84

It helped me get it working right.  I think it also has value to keep something like this in place, though.  Without it, if bisect's not working for a user, there's not much we can do to understand why.  Having a verbose/debug mode will allow them to give us a print out that will hopefully allow us to understand where it went wrong.  Bundler has something similar -- see bundler/bundler#2248.

That said, this is a crappy WIP implementation of such debugging output.  I'm thinking the debug output should be moved into the formatter (or perhaps be an alternate formatter).  And perhaps we shouldn't necessarily use an ENV var to activate it, although I don't necessarily want to add a public API for it either -- it's more just something we can ask users to enable when they report bisect bugs.

Thoughts?